cpp/src/arrow/compute/exec/spilling_util.cc

```diff
@@ -0,0 +1,253 @@
+#include "spilling_util.h"
+
+namespace arrow
+{
+namespace compute
+{
+
+    struct ArrayInfo
+    {
+        int64_t num_children;
+        std::array<std::shared_ptr<Buffer>, 3> bufs;
+        std::array<size_t, 3> sizes;
+        bool has_dict;
+    };
+
+    struct SpillFile::BatchInfo
+    {
+        int64_t start;
+        std::vector<ArrayInfo> arrays;
+    };
+
+#ifdef _WIN32
+#include "windows_compatibility.h"
+
+const FileHandle kInvalidHandle = INVALID_HANDLE_VALUE;
+
+static Result<FileHandle> OpenTemporaryFile()
```

Comment:
We already have some temporary file handling utilities in util/io_util.cc.  Is there a reason you couldn't reuse those?  Also, should this be a temporary directory anyway?  It seems like forcing users to configure and specify a spilling directory might help them to be more aware of what is going on.

Reply:
The only thing in `io_util` that I saw was creating a temporary directory, nothing to do with making temp files.

Reply:
Regarding forcing them to make a spilling directory, ideally I'd like spilling to be completely transparent to them. If they do care, they can always set the environment variable.